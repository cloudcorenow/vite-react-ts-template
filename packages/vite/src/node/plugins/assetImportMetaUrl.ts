import type { Plugin } from '../plugin'
import MagicString from 'magic-string'
import path from 'path'
import { fileToUrl } from './asset'
import type { ResolvedConfig } from '../config'
import {
  blankReplacer,
  multilineCommentsRE,
  singlelineCommentsRE
} from '../utils'
import { htmlTypesRE, scriptRE } from '../optimizer/scan'
import { JS_TYPES_RE } from '../constants'

/**
 * Convert `new URL('./foo.png', import.meta.url)` to its resolved built URL
 *
 * Supports template string with dynamic segments:
 * ```
 * new URL(`./dir/${name}.png`, import.meta.url)
 * // transformed to
 * import.meta.globEager('./dir/**.png')[`./dir/${name}.png`].default
 * ```
 */
export function assetImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
  return {
    name: 'vite:asset-import-meta-url',
    async transform(code, id, options) {
      let inHTML = false
      if (htmlTypesRE.test(id)) {
        inHTML = true
      } else if (JS_TYPES_RE.test(id)) {
        inHTML = false
      } else {
        return
      }

      if (
        options?.ssr ||
        !code.includes('new URL') ||
        !code.includes(`import.meta.url`)
      ) {
        return
      }
      let s: MagicString | undefined

      const transformAssetImportMetaUrl = async (
        snippet: string,
        start: number
      ): Promise<void> => {
        const importMetaUrlRE =
          /\bnew\s+URL\s*\(\s*('[^']+'|"[^"]+"|`[^`]+`)\s*,\s*import\.meta\.url\s*,?\s*\)/g
        const noCommentsCode = snippet
          .replace(multilineCommentsRE, blankReplacer)
          .replace(singlelineCommentsRE, blankReplacer)

        let match: RegExpExecArray | null
        while ((match = importMetaUrlRE.exec(noCommentsCode))) {
          const { 0: exp, 1: rawUrl, index: matchIndex } = match
          const index = start + matchIndex
          s ||= new MagicString(code)

          // potential dynamic template string
          if (rawUrl[0] === '`' && /\$\{/.test(rawUrl)) {
            const ast = this.parse(rawUrl)
            const templateLiteral = (ast as any).body[0].expression
            if (templateLiteral.expressions.length) {
              const pattern = buildGlobPattern(templateLiteral)
              // Note: native import.meta.url is not supported in the baseline
              // target so we use the global location here. It can be
              // window.location or self.location in case it is used in a Web Worker.
              // @see https://developer.mozilla.org/en-US/docs/Web/API/Window/self
              s.overwrite(
                index,
                index + exp.length,
                `new URL(import.meta.globEagerDefault(${JSON.stringify(
                  pattern
                )})[${rawUrl}], self.location)`,
                { contentOnly: true }
              )
              continue
            }
          }

          const url = rawUrl.slice(1, -1)
          const file = path.resolve(path.dirname(id), url)
          // Get final asset URL. Catch error if the file does not exist,
          // in which we can resort to the initial URL and let it resolve in runtime
          const builtUrl = await fileToUrl(file, config, this).catch(() => {
            config.logger.warnOnce(
              `\n${exp} doesn't exist at build time, it will remain unchanged to be resolved at runtime`
            )
            return url
          })
          s.overwrite(
            index,
            index + exp.length,
            `new URL(${JSON.stringify(builtUrl)}, self.location)`,
            { contentOnly: true }
          )
        }
      }

      if (inHTML) {
        let scriptMatch: RegExpExecArray | null = null
        while ((scriptMatch = scriptRE.exec(code))) {
          const { 0: exp, 2: script, index: scriptMatchIndex } = scriptMatch
          const index = exp.indexOf(script) + scriptMatchIndex
          await transformAssetImportMetaUrl(script, index)
        }
      } else {
        await transformAssetImportMetaUrl(code, 0)
      }

      if (s) {
        return {
          code: s.toString(),
          map: config.build.sourcemap ? s.generateMap({ hires: true }) : null
        }
      }
      return null
    }
  }
}

function buildGlobPattern(ast: any) {
  let pattern = ''
  let lastElementIndex = -1
  for (const exp of ast.expressions) {
    for (let i = lastElementIndex + 1; i < ast.quasis.length; i++) {
      const el = ast.quasis[i]
      if (el.end < exp.start) {
        pattern += el.value.raw
        lastElementIndex = i
      }
    }
    pattern += '**'
  }
  for (let i = lastElementIndex + 1; i < ast.quasis.length; i++) {
    pattern += ast.quasis[i].value.raw
  }
  return pattern
}
