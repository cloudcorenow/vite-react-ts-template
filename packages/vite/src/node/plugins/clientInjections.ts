import path from 'node:path'
import type { Plugin } from '../plugin'
import type { ResolvedConfig } from '../config'
import { CLIENT_ENTRY, ENV_ENTRY } from '../constants'
import {
  isObject,
  normalizePath,
  replaceInCode,
  resolveHostname
} from '../utils'

// ids in transform are normalized to unix style
const normalizedClientEntry = normalizePath(CLIENT_ENTRY)
const normalizedEnvEntry = normalizePath(ENV_ENTRY)

/**
 * some values used by the client needs to be dynamically injected by the server
 * @server-only
 */
export function clientInjectionsPlugin(config: ResolvedConfig): Plugin {
  return {
    name: 'vite:client-inject',
    async transform(code, id, options) {
      if (id === normalizedClientEntry || id === normalizedEnvEntry) {
        const resolvedServerHostname = (
          await resolveHostname(config.server.host)
        ).name
        const resolvedServerPort = config.server.port!
        const devBase = config.base

        const serverHost = `${resolvedServerHostname}:${resolvedServerPort}${devBase}`

        let hmrConfig = config.server.hmr
        hmrConfig = isObject(hmrConfig) ? hmrConfig : undefined
        const host = hmrConfig?.host || null
        const protocol = hmrConfig?.protocol || null
        const timeout = hmrConfig?.timeout || 30000
        const overlay = hmrConfig?.overlay !== false

        // hmr.clientPort -> hmr.port
        // -> (24678 if middleware mode) -> new URL(import.meta.url).port
        let port = hmrConfig?.clientPort || hmrConfig?.port || null
        if (config.server.middlewareMode) {
          port ||= 24678
        }

        let directTarget = hmrConfig?.host || resolvedServerHostname
        directTarget += `:${hmrConfig?.port || resolvedServerPort}`
        directTarget += devBase

        let hmrBase = devBase
        if (hmrConfig?.path) {
          hmrBase = path.posix.join(hmrBase, hmrConfig.path)
        }

        const replacements: Record<string, string> = {
          __MODE__: JSON.stringify(config.mode),
          __BASE__: JSON.stringify(devBase),
          __DEFINES__: serializeDefine(config.define || {}),
          __SERVER_HOST__: JSON.stringify(serverHost),
          __HMR_PROTOCOL__: JSON.stringify(protocol),
          __HMR_HOSTNAME__: JSON.stringify(host),
          __HMR_PORT__: JSON.stringify(port),
          __HMR_DIRECT_TARGET__: JSON.stringify(directTarget),
          __HMR_BASE__: JSON.stringify(hmrBase),
          __HMR_TIMEOUT__: JSON.stringify(timeout),
          __HMR_ENABLE_OVERLAY__: JSON.stringify(overlay)
        }
        const pattern = new RegExp(Object.keys(replacements).join('|'), 'gu')
        return replaceInCode(code, pattern, replacements)?.toString() ?? null
      } else if (!options?.ssr) {
        // replace process.env.NODE_ENV instead of defining a global
        // for it to avoid shimming a `process` object during dev,
        // avoiding inconsistencies between dev and build
        return (
          replaceInCode(
            code,
            /\bprocess\.env\.NODE_ENV\b/g,
            config.define?.['process.env.NODE_ENV'] ||
              JSON.stringify(process.env.NODE_ENV || config.mode)
          )?.toString() ?? null
        )
      }
      return null
    }
  }
}

function serializeDefine(define: Record<string, any>): string {
  let res = `{`
  for (const key in define) {
    const val = define[key]
    res += `${JSON.stringify(key)}: ${
      typeof val === 'string' ? `(${val})` : JSON.stringify(val)
    }, `
  }
  return res + `}`
}
