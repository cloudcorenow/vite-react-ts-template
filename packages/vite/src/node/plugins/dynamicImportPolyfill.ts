import { ResolvedConfig } from '..'
import { Plugin } from '../plugin'

export const polyfillId = 'vite/dynamic-import-polyfill'

export function dynamicImportPolyfillPlugin(config: ResolvedConfig): Plugin {
  return {
    name: 'vite:dynamic-import-polyfill',
    resolveId(id) {
      if (id === polyfillId) {
        return id
      }
    },
    load(id) {
      if (id === polyfillId) {
        config.logger.warn(
          `\nVite's dynamic import polyfill has been removed, stop importing 'vite/dynamic-import-polyfill'`
        )
        return ''
      }
    }
  }
}
