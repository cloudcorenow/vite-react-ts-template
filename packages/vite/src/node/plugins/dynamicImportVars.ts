import path from 'path'
import MagicString from 'magic-string'
import { init, parse as parseImports } from 'es-module-lexer'
import type { ImportSpecifier } from 'es-module-lexer'
import type { Plugin } from '../plugin'
import type { ResolvedConfig } from '../config'
import { parseRequest } from '../utils'
import { parse as parseJS } from 'acorn'
import { createFilter } from '@rollup/pluginutils'
import { dynamicImportToGlob } from '@rollup/plugin-dynamic-import-vars'
import type { TransformGlobImportResult } from './importMetaGlob'
import { transformGlobImport } from './importMetaGlob'

interface DynamicImportRequest {
  as?: 'raw'
}

interface DynamicImportPattern {
  globParams: DynamicImportRequest | null
  userPattern: string
  rawPattern: string
}

const dynamicImportHelper = (glob: Record<string, any>, path: string) => {
  const v = glob[path]
  if (v) {
    return typeof v === 'function' ? v() : Promise.resolve(v)
  }
  return new Promise((_, reject) => {
    ;(typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
      reject.bind(null, new Error('Unknown variable dynamic import: ' + path))
    )
  })
}
export const dynamicImportHelperId = '/@vite/dynamic-import-helper'

export function parseDynamicImportPattern(
  strings: string
): DynamicImportPattern | null {
  const filename = strings.slice(1, -1)
  const rawQuery = parseRequest(filename)
  let globParams: DynamicImportRequest | null = null
  const ast = (
    parseJS(strings, {
      ecmaVersion: 'latest',
      sourceType: 'module'
    }) as any
  ).body[0].expression

  const userPatternQuery = dynamicImportToGlob(ast, filename)
  if (!userPatternQuery) {
    return null
  }

  const [userPattern] = userPatternQuery.split('?', 2)
  const [rawPattern] = filename.split('?', 2)

  if (rawQuery?.raw !== undefined) {
    globParams = { as: 'raw' }
  }

  return {
    globParams,
    userPattern,
    rawPattern
  }
}

export async function transformDynamicImportGlob(
  source: string,
  root: string,
  importer: string,
  start: number,
  end: number,
  resolve: (url: string, importer?: string) => Promise<string | undefined>
): Promise<{
  glob: TransformGlobImportResult
  pattern: string
  rawPattern: string
} | null> {
  let fileName = source.slice(start, end)

  if (fileName[1] !== '.' && fileName[1] !== '/' && resolve) {
    const resolvedFileName = await resolve(fileName.slice(1, -1), importer)
    if (!resolvedFileName) {
      return null
    }
    const relativeFileName = path.posix.relative(
      path.dirname(importer),
      resolvedFileName
    )
    fileName = `\`${relativeFileName}\``
  }

  const dynamicImportPattern = parseDynamicImportPattern(fileName)
  if (!dynamicImportPattern) {
    return null
  }
  const { globParams, rawPattern, userPattern } = dynamicImportPattern
  const params = globParams ? `, ${JSON.stringify(globParams)}` : ''
  const exp = `import.meta.glob(${JSON.stringify(userPattern)}${params})`
  const glob = await transformGlobImport(
    exp,
    importer,
    root,
    resolve,
    false,
    false
  )

  return {
    rawPattern,
    pattern: userPattern,
    glob: glob!
  }
}

export function dynamicImportHelperPlugin(): Plugin {
  return {
    name: 'vite:dynamic-import-helper',
    resolveId(id) {
      if (id === dynamicImportHelperId) {
        return id
      }
    },

    load(id) {
      if (id === dynamicImportHelperId) {
        return 'export default' + dynamicImportHelper.toString()
      }
    }
  }
}

export function dynamicImportVarsPlugin(config: ResolvedConfig): Plugin {
  const resolve = config.createResolver({
    preferRelative: true,
    tryIndex: false,
    extensions: []
  })
  const { include, exclude, warnOnError } =
    config.build.dynamicImportVarsOptions
  const filter = createFilter(include, exclude)

  return {
    name: 'vite:dynamic-import-vars',

    async transform(source, importer) {
      if (!filter(importer)) {
        return
      }

      await init

      let imports: readonly ImportSpecifier[] = []
      try {
        imports = parseImports(source)[0]
      } catch (e: any) {
        this.error(e, e.idx)
      }

      if (!imports.length) {
        return null
      }

      let s: MagicString | undefined
      let needDynamicImportHelper = false

      for (let index = 0; index < imports.length; index++) {
        const {
          s: start,
          e: end,
          ss: expStart,
          se: expEnd,
          d: dynamicIndex
        } = imports[index]

        if (dynamicIndex === -1 || source[start] !== '`') {
          continue
        }

        s ||= new MagicString(source)
        let result
        try {
          result = await transformDynamicImportGlob(
            source,
            config.root,
            importer,
            start,
            end,
            resolve
          )
        } catch (error) {
          if (warnOnError) {
            this.warn(error)
          } else {
            this.error(error)
          }
        }

        if (!result) {
          continue
        }

        const { rawPattern, glob } = result

        needDynamicImportHelper = true
        s.overwrite(
          expStart,
          expEnd,
          `__variableDynamicImportRuntimeHelper(${glob.s.toString()}, \`${rawPattern}\`)`
        )
      }

      if (s) {
        if (needDynamicImportHelper) {
          s.prepend(
            `import __variableDynamicImportRuntimeHelper from "${dynamicImportHelperId}";`
          )
        }
        return {
          code: s.toString(),
          map: config.build.sourcemap ? s.generateMap({ hires: true }) : null
        }
      }
    }
  }
}
