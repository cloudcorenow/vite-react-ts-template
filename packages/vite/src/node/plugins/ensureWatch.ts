import type { Plugin } from '../plugin'
import { cleanUrl } from '../utils'

/**
 * plugin to ensure rollup can watch correctly.
 */
export function ensureWatchPlugin(): Plugin {
  return {
    name: 'vite:ensure-watch',
    load(id) {
      if (id.includes('?')) {
        this.addWatchFile(cleanUrl(id))
      }
      return null
    }
  }
}
