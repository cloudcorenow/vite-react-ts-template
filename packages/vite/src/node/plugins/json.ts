/**
 * https://github.com/rollup/plugins/blob/master/packages/json/src/index.js
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file at
 * https://github.com/rollup/plugins/blob/master/LICENSE
 */

import { dataToEsm, makeLegalIdentifier } from '@rollup/pluginutils'
import { SPECIAL_QUERY_RE } from '../constants'
import type { Plugin } from '../plugin'
import { stripBomTag } from '../utils'

export interface JsonOptions {
  /**
   * Generate a named export for every property of the JSON object
   * @default true
   */
  namedExports?: boolean
  /**
   * Generate performant output as JSON.parse("stringified").
   *
   * When set to 'auto', the data will be stringified only if the data is bigger than 10kB.
   * @default false
   */
  stringify?: boolean | 'auto'
}

// Custom json filter for vite
const jsonExtRE = /\.json(?:$|\?)(?!commonjs-(?:proxy|external))/

const jsonLangs = `\\.(?:json|json5)(?:$|\\?)`
const jsonLangRE = new RegExp(jsonLangs)
export const isJSONRequest = (request: string): boolean =>
  jsonLangRE.test(request)

export function jsonPlugin(
  options: JsonOptions = {},
  isBuild: boolean,
): Plugin {
  return {
    name: 'vite:json',

    transform(json, id) {
      if (!jsonExtRE.test(id)) return null
      if (SPECIAL_QUERY_RE.test(id)) return null

      json = stripBomTag(json)

      try {
        const stringify =
          options.stringify === true ||
          // use 10kB as a threshold
          // https://v8.dev/blog/cost-of-javascript-2019#:~:text=A%20good%20rule%20of%20thumb%20is%20to%20apply%20this%20technique%20for%20objects%20of%2010%20kB%20or%20larger
          (options.stringify === 'auto' && json.length >= 10 * 1000)

        const parsed = JSON.parse(json)

        if (stringify) {
          const contentCode = isBuild
            ? // during build, parse then double-stringify to remove all
              // unnecessary whitespaces to reduce bundle size.
              `JSON.parse(${JSON.stringify(JSON.stringify(parsed))})`
            : `JSON.parse(${JSON.stringify(json)})`

          let code: string
          if (options.namedExports) {
            let defaultKey = 'default_'
            const keys = Object.keys(parsed)
            const keysSet = new Set(keys)
            while (keysSet.has(defaultKey)) {
              defaultKey += '_'
            }

            code = `const ${defaultKey} = ${contentCode};\nexport default default_;\n`
            for (const key of keys) {
              if (key === makeLegalIdentifier(key)) {
                code += `export const ${key} = ${defaultKey}.${key};\n`
              }
            }
          } else {
            code = `export default ${contentCode}`
          }

          return {
            code,
            map: { mappings: '' },
          }
        }

        return {
          code: dataToEsm(parsed, {
            preferConst: true,
            namedExports: options.namedExports,
          }),
          map: { mappings: '' },
        }
      } catch (e) {
        const position = extractJsonErrorPosition(e.message, json.length)
        const msg = position
          ? `, invalid JSON syntax found at position ${position}`
          : `.`
        this.error(`Failed to parse JSON file` + msg, position)
      }
    },
  }
}

export function extractJsonErrorPosition(
  errorMessage: string,
  inputLength: number,
): number | undefined {
  if (errorMessage.startsWith('Unexpected end of JSON input')) {
    return inputLength - 1
  }

  const errorMessageList = /at position (\d+)/.exec(errorMessage)
  return errorMessageList
    ? Math.max(parseInt(errorMessageList[1], 10) - 1, 0)
    : undefined
}
