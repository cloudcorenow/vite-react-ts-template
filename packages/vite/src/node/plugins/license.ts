import fs from 'node:fs'
import path from 'node:path'
import type { Plugin } from '../plugin'
import { isInNodeModules, sortObjectKeys } from '../utils'
import type { PackageCache } from '../packages'
import { findNearestMainPackageData } from '../packages'

export interface LicenseEntry {
  /**
   * Package name
   */
  name: string
  /**
   * Package version
   */
  version: string
  /**
   * SPDX license identifier (from package.json "license" field)
   */
  identifier?: string
  /**
   * License file text
   */
  text?: string
}

// https://github.com/npm/npm-packlist/blob/53b2a4f42b7fef0f63e8f26a3ea4692e23a58fed/lib/index.js#L284-L286
const licenseFiles = [/^license/i, /^licence/i, /^copying/i]

export interface ManifestChunk {
  src?: string
  file: string
  css?: string[]
  assets?: string[]
  isEntry?: boolean
  name?: string
  isDynamicEntry?: boolean
  imports?: string[]
  dynamicImports?: string[]
}

export function licensePlugin(): Plugin {
  return {
    name: 'vite:manifest',

    async generateBundle(_, bundle) {
      const licenseOption = this.environment.config.build.license
      if (licenseOption === false) return

      const packageCache: PackageCache = new Map()
      // Track license via a key to its license entry.
      // A key consists of "name@version" of a package.
      const licenses: Record<string, LicenseEntry> = {}

      for (const file in bundle) {
        const chunk = bundle[file]
        if (chunk.type === 'asset') continue

        for (const moduleId of chunk.moduleIds) {
          if (moduleId.startsWith('\0') || !isInNodeModules(moduleId)) continue

          // Find the dependency package.json
          const pkgData = findNearestMainPackageData(
            path.dirname(moduleId),
            packageCache,
          )
          if (!pkgData) continue

          // Grab the package.json keys and check if already exists in the licenses
          const { name, version = '0.0.0', license } = pkgData.data
          const key = `${name}@${version}`
          if (licenses[key]) continue

          // If not, create a new license entry
          const entry: LicenseEntry = { name, version }
          if (license) {
            entry.identifier = license.trim()
          }
          const licenseFile = findLicenseFile(pkgData.dir)
          if (licenseFile) {
            entry.text = fs.readFileSync(licenseFile, 'utf-8').trim()
          }
          licenses[key] = entry
        }
      }

      const licenseEntries = Object.values(sortObjectKeys(licenses))

      // Emit as a JSON file
      if (
        typeof licenseOption === 'string' &&
        licenseOption.endsWith('.json')
      ) {
        this.emitFile({
          fileName: licenseOption,
          type: 'asset',
          source: JSON.stringify(licenseEntries, null, 2),
        })
        return
      }

      // Emit a license file as markdown
      const markdown = licenseEntryToMarkdown(licenseEntries)
      this.emitFile({
        fileName:
          typeof licenseOption === 'string'
            ? licenseOption
            : '.vite/license.md',
        type: 'asset',
        source: markdown,
      })
    },
  }
}

function licenseEntryToMarkdown(licenses: LicenseEntry[]) {
  if (licenses.length === 0) {
    return `\
# Licenses

The app does not bundle any dependencies with licenses.
`
  }

  let text = `\
# Licenses

The app bundles dependencies which contains the following licenses:
`
  for (const license of licenses) {
    const nameAndVersionText = `${license.name} - ${license.version}`
    const identifierText = license.identifier ? ` (${license.identifier})` : ''

    text += `\n## ${nameAndVersionText}${identifierText}\n`
    if (license.text) {
      text += `\n${license.text}\n`
    }
  }
  return text
}

function findLicenseFile(pkgDir: string) {
  const files = fs.readdirSync(pkgDir)
  const matchedFile = files.find((file) =>
    licenseFiles.some((re) => re.test(file)),
  )
  if (matchedFile) {
    return path.join(pkgDir, matchedFile)
  }
}
