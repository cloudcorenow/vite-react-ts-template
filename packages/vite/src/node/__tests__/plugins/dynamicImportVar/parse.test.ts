import { describe, expect, it } from 'vitest'
import { transformDynamicImport } from '../../../plugins/dynamicImportVars'
import { resolve } from 'path'

async function run(input: string) {
  const { glob, rawPattern } = await transformDynamicImport(
    input,
    resolve(__dirname),
    resolve(__dirname, 'index.js'),
    (id) => id.replace('@', resolve(__dirname, './mods/'))
  )
  return `__variableDynamicImportRuntimeHelper(${glob.s.toString()}, \`${rawPattern}\`)`
}

describe('parse positives', () => {
  it('basic', async () => {
    expect(await run('`./mods/${base}.js`')).toMatchSnapshot()
  })

  it('alias path', async () => {
    expect(await run('`@/${base}.js`')).toMatchSnapshot()
  })

  it('with query raw', async () => {
    expect(await run('`./mods/${base}.js?raw`')).toMatchSnapshot()
  })

  it('with query url', async () => {
    expect(await run('`./mods/${base}.js?url`')).toMatchSnapshot()
  })
})
