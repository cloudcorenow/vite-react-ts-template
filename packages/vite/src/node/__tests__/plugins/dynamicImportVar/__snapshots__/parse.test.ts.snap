// Vitest Snapshot v1

exports[`parse positives > alias path 1`] = `
"__variableDynamicImportRuntimeHelper({
\\"./mods/hello.js\\": () => import(\\"./mods/hello.js\\"),
\\"./mods/hi.js\\": () => import(\\"./mods/hi.js\\")
}, \`./mods/\${base}.js\`)"
`;

exports[`parse positives > basic 1`] = `
"__variableDynamicImportRuntimeHelper({
\\"./mods/hello.js\\": () => import(\\"./mods/hello.js\\"),
\\"./mods/hi.js\\": () => import(\\"./mods/hi.js\\")
}, \`./mods/\${base}.js\`)"
`;

exports[`parse positives > with query raw 1`] = `
"__variableDynamicImportRuntimeHelper({
\\"./mods/hello.js\\": () => import(\\"./mods/hello.js?raw\\"),
\\"./mods/hi.js\\": () => import(\\"./mods/hi.js?raw\\")
}, \`./mods/\${base}.js\`)"
`;

exports[`parse positives > with query url 1`] = `
"__variableDynamicImportRuntimeHelper({
\\"./mods/hello.js\\": () => import(\\"./mods/hello.js\\"),
\\"./mods/hi.js\\": () => import(\\"./mods/hi.js\\")
}, \`./mods/\${base}.js\`)"
`;
